File: fft_rust/src/fft_loop.rs
use num_complex::Complex64;

fn bit_reverse(mut x: usize, log2_n: usize) -> usize {
    let mut n = 0;
    for _i in 0..log2_n {
        n <<= 1;
        n |= x & 1;
        x >>= 1;
    }

    n
}

// is_reverseがtrueなら逆フーリエ変換を実行(規格化も行う)
fn fft_loop_ref(array: &mut Vec<Complex64>, is_reverse: bool) {
    let n = array.len();
    let log2_n = (n as f64).log2().floor() as usize;
    let len_arry = Complex64::new(n as f64, 0.0);

    // bit reversal of the given array
    let buff = array.clone();
    for i in 0..n {
        let idx_rev = bit_reverse(i, log2_n);
        array[i] = buff[idx_rev];
        if is_reverse {
            array[i] /= len_arry;
        }
    }

    for s in 1..=log2_n {
        let m = 1 << s;
        let m_2 = m >> 1;
        let mut w = Complex64::new(1.0, 0.0);

        let mut w_m = Complex64::new(0.0, std::f64::consts::PI / m_2 as f64).exp();
        if is_reverse {
            w_m = w_m.conj();
        }

        for j in 0..m_2 {
            let mut k = j;
            while k < n {
                let t = w * array[k + m_2];
                let u = array[k];

                array[k] = u + t;
                array[k + m_2] = u - t;

                k += m;
            }
            w *= w_m;
        }
    }
}

fn main() {
    let mut array: Vec<Complex64> = vec![1.0, 2.0, 3.0, 4.0]
        .iter()
        .map(|x| Complex64::new(*x, 0.0))
        .collect();

    // フーリエ変換
    fft_loop_ref(&mut array, false);
    println!("array: {:?}", array);

    // 逆フーリエ変換
    fft_loop_ref(&mut array, true);
    println!("array: {:?}", array);

    let mut array: Vec<Complex64> = vec![1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0]
        .iter()
        .map(|x| Complex64::new(*x, 0.0))
        .collect();

    // フーリエ変換
    fft_loop_ref(&mut array, false);
    println!("array:\n {:?}", array);

    // 逆フーリエ変換
    fft_loop_ref(&mut array, true);
    println!("array:\n {:?}", array);
}

File: fft_cpp/fft_loop.cpp
#include <array>
#include <iostream>
#include <vector>
#include <complex>

/*
 Iterative Fast Fourier Transformation for polynomial multiplication:

https://www.geeksforgeeks.org/iterative-fast-fourier-transformation-polynomial-multiplication/
 */

void show_result(std::vector<std::complex<double>>& result) {
    for(auto& v: result) {
        std::cout << v << ",";
    }
    std::cout << std::endl;
}

std::size_t bit_reverse(std::size_t x, std::size_t log2_n) {
    std::size_t n = 0;
    for(std::size_t i = 0; i < log2_n; ++i) {
        n <<= 1;
        n |= (x & 1);
        x >>= 1;
    }

    return n;
}

// is_reverseがtrueなら逆フーリエ変換を実行(規格化も行う)
void fft_loop_ref(std::vector<std::complex<double>>& array, bool is_reverse) {
    auto n = array.size();
    auto log2_n = std::log2(n);
    auto len_array = std::complex((double)array.size(), 0.0);

    // bit reversal of th given array
    auto buff = array;
    for(std::size_t i = 0; i < n; ++i) {
        auto idx_rev = bit_reverse(i, log2_n);
        array[i] = buff[idx_rev];
        if(is_reverse) {
            array[i] /= len_array;
        }
    }

    for(std::size_t s = 1; s <= log2_n; ++s) {
        auto m = 1 << s;
        auto m_2 = m >> 1;
        std::complex<double> w(1.0, 0.0);

        std::complex<double> w_m = std::exp(std::complex<double>(0, M_PI / m_2));
        if(is_reverse) {
            w_m = std::conj(w_m);
        }

        for(std::size_t j = 0; j < m_2; ++j) {
            for(std::size_t k = j; k < n; k += m) {

                auto t = w * array[k + m_2];
                auto u = array[k];

                array[k] = u + t;
                array[k + m_2] = u - t;
            }
            w *= w_m;
        } 
    }
}

int main(void) {
    std::vector<std::complex<double>> array = {1.0, 2.0, 3.0, 4.0};
    std::vector<std::complex<double>> array_res(4);

    // フーリエ変換
    auto res_2 = array;
    fft_loop_ref(res_2, false);
    show_result(res_2);

    // 逆フーリエ変換
    fft_loop_ref(res_2, true);
    show_result(res_2);
}
